// @vtex/api/lib/service/logger/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const error_1 = require("../../utils/error");
const linked = !!process.env.VTEX_APP_LINK;
const app = process.env.VTEX_APP_ID;
const EMPTY_MESSAGE = 'Logger.log was called with null or undefined message';
const appName = process.env.VTEX_APP_NAME;
const vendor = process.env.VTEX_APP_VENDOR;
var LogLevel;
(function (LogLevel) {
    LogLevel["Debug"] = "debug";
    LogLevel["Info"] = "info";
    LogLevel["Warn"] = "warn";
    LogLevel["Error"] = "error";
})(LogLevel = exports.LogLevel || (exports.LogLevel = {}));
class Logger {
    constructor(ctx) {
        this.executed = false;
        this.debug = (message) => this.log(message, LogLevel.Debug);
        this.info = (message) => this.log(message, LogLevel.Info);
        this.warn = (warning) => this.log(warning, LogLevel.Warn);
        this.error = (error) => this.log(error, LogLevel.Error);
        this.log = (message, level) => {
            const data = message ? error_1.cleanError(message) : EMPTY_MESSAGE;
            /* tslint:disable:object-literal-sort-keys */
            const inflatedLog = {
                __VTEX_IO_LOG: true,
                level,
                app,
                account: this.account,
                workspace: this.workspace,
                production: this.production,
                data,
                operationId: this.operationId,
                requestId: this.requestId,
            };
            if (this.isThirdPartyApp()) {
                Object.assign(inflatedLog, {
                    __SKIDDER_TOPIC_1: `skidder.acc.${this.account}`,
                    __SKIDDER_TOPIC_2: `skidder.app.${this.account}.${vendor}.${appName}`,
                });
            }
            console.log(JSON.stringify(inflatedLog));
            // Warn the developer how to retrieve the error in splunk
            this.logSplunkQuery();
        };
        /**
         * Logs splunk query so the developer can search for the errors in splunk.
         * This function runs only once in the lifetime of the Logger class so we
         * don't mess up with the developer's terminal
         */
        this.logSplunkQuery = () => {
            if (!this.executed && linked) {
                console.log(`Try this query at Splunk to retrieve error log: 'index=io_vtex_logs app="${app}" account=${this.account} workspace=${this.workspace} level=error OR level=warn'`);
                this.executed = true;
            }
        };
        this.account = ctx.account;
        this.workspace = ctx.workspace;
        this.requestId = ctx.requestId;
        this.operationId = ctx.operationId;
        this.production = ctx.production;
    }
    isThirdPartyApp() {
        return 'vtex' !== vendor && 'gocommerce' !== vendor;
    }
}
exports.Logger = Logger;
